crates/recursion/circuit-v2/src/merkle_tree.rs

```diff
@@ -0,0 +1,191 @@
+use std::fmt::Debug;
+
+use itertools::Itertools;
+use p3_field::Field;
+use p3_util::{reverse_bits_len, reverse_slice_index_bits};
+use sp1_core_machine::utils::log2_strict_usize;
+use sp1_recursion_compiler::ir::Builder;
+
+use crate::{
+    hash::{FieldHasher, FieldHasherVariable},
+    CircuitConfig,
+};
+
+#[derive(Debug, Clone)]
+pub struct MerkleTree<F: Field, HV: FieldHasher<F>> {
+    /// The height of the tree, not counting the root layer. This is the same as the logarithm of the
+    /// number of leaves.
+    pub height: usize,
+
+    /// All the layers but the root. If there are `n` leaves where `n` is a power of 2, there are
+    /// `2n - 2` elements in this vector. The leaves are at the beginning of the vector.
+    pub digest_layers: Vec<HV::Digest>,
+}
+pub struct VcsError;
+
+impl Debug for VcsError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "VcsError")
+    }
+}
+
+impl<F: Field, HV: FieldHasher<F>> MerkleTree<F, HV> {
+    pub fn commit(leaves: Vec<HV::Digest>) -> (HV::Digest, Self) {
+        assert!(!leaves.is_empty());
+        let new_len = leaves.len().next_power_of_two();
+        let height = log2_strict_usize(new_len);
+
+        // Pre-allocate the vector.
+        let mut digest_layers = Vec::with_capacity(2 * new_len - 2);
+
+        // If `leaves.len()` is not a power of 2, we pad the leaves with default values.
+        let mut last_layer = leaves.clone();
```

Comment:
Why cloning here? You can just change the signature of the function to `leaves: mut Vec<HV::Digest>`. It's ok because you are taking ownership

Reply:
No good reason. Removed it.